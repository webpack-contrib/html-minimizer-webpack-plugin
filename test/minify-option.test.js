import HtmlMinimizerPlugin from "../src/index";

import {
  compile,
  getCompiler,
  getErrors,
  getWarnings,
  readAssets,
} from "./helpers";

describe('"minify" option', () => {
  it("should work minify function", async () => {
    const testHtmlId = "./simple.html";
    const compiler = getCompiler(testHtmlId);

    new HtmlMinimizerPlugin({
      minimizerOptions: {
        collapseWhitespace: true,
      },
      minify: (data, minimizerOptions) => {
        // eslint-disable-next-line global-require
        const htmlMinifier = require("html-minifier-terser");
        const [[, input]] = Object.entries(data);

        return htmlMinifier.minify(input, minimizerOptions);
      },
    }).apply(compiler);

    const stats = await compile(compiler);

    expect(readAssets(compiler, stats, /\.html$/i)).toMatchSnapshot("assets");
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it("should work if minify is array && minimizerOptions is object", async () => {
    const testHtmlId = "./simple.html";
    const compiler = getCompiler(testHtmlId);

    new HtmlMinimizerPlugin({
      minimizerOptions: {
        collapseWhitespace: true,
      },
      minify: [
        HtmlMinimizerPlugin.htmlMinifierTerser,
        async (data, minimizerOptions) => {
          const [[, input]] = Object.entries(data);

          return `${input}<div>Second function: ${minimizerOptions.collapseWhitespace}</div>`;
        },
        async (data, minimizerOptions) => {
          const [[name, input]] = Object.entries(data);
          return `${name} - ${input}<div>Third function: ${minimizerOptions.collapseWhitespace}</div>`;
        },
      ],
    }).apply(compiler);

    const stats = await compile(compiler);

    expect(readAssets(compiler, stats, /\.html$/i)).toMatchSnapshot("assets");
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it("should work if minify is array && minimizerOptions is array", async () => {
    const testHtmlId = "./simple.html";
    const compiler = getCompiler(testHtmlId);

    new HtmlMinimizerPlugin({
      minimizerOptions: [
        {
          collapseWhitespace: true,
        },
        {
          test: "text from option",
        },
      ],
      minify: [
        HtmlMinimizerPlugin.htmlMinifierTerser,
        async (data, minimizerOptions) => {
          const [[name, input]] = Object.entries(data);
          return `${name} - ${input}:<div>Second function: ${minimizerOptions.test}</div>`;
        },
      ],
    }).apply(compiler);

    const stats = await compile(compiler);

    expect(readAssets(compiler, stats, /\.html$/i)).toMatchSnapshot("assets");
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it("should work if minify function return object", async () => {
    const testHtmlId = "./simple.html";
    const compiler = getCompiler(testHtmlId);

    new HtmlMinimizerPlugin({
      minimizerOptions: [
        {
          collapseWhitespace: true,
        },
        {
          test: "text from option",
        },
      ],
      minify: [
        HtmlMinimizerPlugin.htmlMinifierTerser,
        async (data, minimizerOptions) => {
          const [[name, input]] = Object.entries(data);
          return {
            code: `${name} - ${input}:<div>Second function: ${minimizerOptions.test}</div>`,
          };
        },
      ],
    }).apply(compiler);

    const stats = await compile(compiler);

    expect(readAssets(compiler, stats, /\.html$/i)).toMatchSnapshot("assets");
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it("should minimize code and emit warning", async () => {
    const testHtmlId = "./simple.html";
    const compiler = getCompiler(testHtmlId);

    new HtmlMinimizerPlugin({
      minimizerOptions: [
        {
          collapseWhitespace: true,
        },
      ],
      minify: [
        HtmlMinimizerPlugin.htmlMinifierTerser,
        async (data) => {
          const [[, input]] = Object.entries(data);
          return {
            code: input,
            warnings: [
              "string warning",
              new Error("test warning"),
              {
                message: "object warning",
              },
            ],
          };
        },
      ],
    }).apply(compiler);

    const stats = await compile(compiler);

    expect(readAssets(compiler, stats, /\.html$/i)).toMatchSnapshot("assets");
    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });

  it("should emit error", async () => {
    const testHtmlId = "./simple.html";
    const compiler = getCompiler(testHtmlId);

    new HtmlMinimizerPlugin({
      minimizerOptions: [
        {
          collapseWhitespace: true,
        },
      ],
      minify: [
        HtmlMinimizerPlugin.htmlMinifierTerser,
        async (data) => {
          const [[, input]] = Object.entries(data);
          return {
            code: input,
            errors: [
              "string error",
              new Error("test error"),
              {
                message: "object error",
              },
            ],
          };
        },
      ],
    }).apply(compiler);

    const stats = await compile(compiler);

    expect(getErrors(stats)).toMatchSnapshot("errors");
    expect(getWarnings(stats)).toMatchSnapshot("warnings");
  });
});
