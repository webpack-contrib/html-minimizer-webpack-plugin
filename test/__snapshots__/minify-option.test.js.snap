// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"minify" option should emit error: errors 1`] = `
Array [
  "Error: simple.html from Html Minimizer plugin
object error",
  "Error: simple.html from Html Minimizer plugin
string error",
  "Error: simple.html from Html Minimizer plugin
test error",
]
`;

exports[`"minify" option should emit error: warnings 1`] = `Array []`;

exports[`"minify" option should minimize code and emit warning: assets 1`] = `
Object {
  "simple.html": "<!doctype html> <html lang=\\"en\\"> <head> <meta charset=\\"UTF-8\\"> <meta name=\\"viewport\\" content=\\"width=device-width,user-scalable=no,initial-scale=1,maximum-scale=1,minimum-scale=1\\"> <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\"> <title>Document</title> </head> <body> <h1>My First Heading</h1> <p>My first paragraph.</p> <h2>An Unordered HTML List</h2> <ul> <li>Coffee</li> <li>Tea</li> <li>Milk</li> </ul> <h2>An Ordered HTML List</h2> <ol> <li>Coffee</li> <li>Tea</li> <li>Milk</li> </ol> </body> </html> ",
}
`;

exports[`"minify" option should minimize code and emit warning: errors 1`] = `Array []`;

exports[`"minify" option should minimize code and emit warning: warnings 1`] = `
Array [
  "Warning: object warning",
  "Warning: string warning",
  "Warning: test warning",
]
`;

exports[`"minify" option should work if minify function return object: assets 1`] = `
Object {
  "simple.html": "simple.html - <!doctype html> <html lang=\\"en\\"> <head> <meta charset=\\"UTF-8\\"> <meta name=\\"viewport\\" content=\\"width=device-width,user-scalable=no,initial-scale=1,maximum-scale=1,minimum-scale=1\\"> <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\"> <title>Document</title> </head> <body> <h1>My First Heading</h1> <p>My first paragraph.</p> <h2>An Unordered HTML List</h2> <ul> <li>Coffee</li> <li>Tea</li> <li>Milk</li> </ul> <h2>An Ordered HTML List</h2> <ol> <li>Coffee</li> <li>Tea</li> <li>Milk</li> </ol> </body> </html> :<div>Second function: text from option</div>",
}
`;

exports[`"minify" option should work if minify function return object: errors 1`] = `Array []`;

exports[`"minify" option should work if minify function return object: warnings 1`] = `Array []`;

exports[`"minify" option should work if minify is array && minimizerOptions is array: assets 1`] = `
Object {
  "simple.html": "simple.html - <!doctype html> <html lang=\\"en\\"> <head> <meta charset=\\"UTF-8\\"> <meta name=\\"viewport\\" content=\\"width=device-width,user-scalable=no,initial-scale=1,maximum-scale=1,minimum-scale=1\\"> <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\"> <title>Document</title> </head> <body> <h1>My First Heading</h1> <p>My first paragraph.</p> <h2>An Unordered HTML List</h2> <ul> <li>Coffee</li> <li>Tea</li> <li>Milk</li> </ul> <h2>An Ordered HTML List</h2> <ol> <li>Coffee</li> <li>Tea</li> <li>Milk</li> </ol> </body> </html> :<div>Second function: text from option</div>",
}
`;

exports[`"minify" option should work if minify is array && minimizerOptions is array: errors 1`] = `Array []`;

exports[`"minify" option should work if minify is array && minimizerOptions is array: warnings 1`] = `Array []`;

exports[`"minify" option should work if minify is array && minimizerOptions is object: assets 1`] = `
Object {
  "simple.html": "simple.html - <!doctype html> <html lang=\\"en\\"> <head> <meta charset=\\"UTF-8\\"> <meta name=\\"viewport\\" content=\\"width=device-width,user-scalable=no,initial-scale=1,maximum-scale=1,minimum-scale=1\\"> <meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\"> <title>Document</title> </head> <body> <h1>My First Heading</h1> <p>My first paragraph.</p> <h2>An Unordered HTML List</h2> <ul> <li>Coffee</li> <li>Tea</li> <li>Milk</li> </ul> <h2>An Ordered HTML List</h2> <ol> <li>Coffee</li> <li>Tea</li> <li>Milk</li> </ol> </body> </html> <div>Second function: true</div><div>Third function: true</div>",
}
`;

exports[`"minify" option should work if minify is array && minimizerOptions is object: errors 1`] = `Array []`;

exports[`"minify" option should work if minify is array && minimizerOptions is object: warnings 1`] = `Array []`;

exports[`"minify" option should work minify function: assets 1`] = `
Object {
  "simple.html": "<!doctype html><html lang=\\"en\\"><head><meta charset=\\"UTF-8\\"><meta name=\\"viewport\\" content=\\"width=device-width,user-scalable=no,initial-scale=1,maximum-scale=1,minimum-scale=1\\"><meta http-equiv=\\"X-UA-Compatible\\" content=\\"ie=edge\\"><title>Document</title></head><body><h1>My First Heading</h1><p>My first paragraph.</p><h2>An Unordered HTML List</h2><ul><li>Coffee</li><li>Tea</li><li>Milk</li></ul><h2>An Ordered HTML List</h2><ol><li>Coffee</li><li>Tea</li><li>Milk</li></ol></body></html>",
}
`;

exports[`"minify" option should work minify function: errors 1`] = `Array []`;

exports[`"minify" option should work minify function: warnings 1`] = `Array []`;
